server/src/internalClusterTest/java/org/elasticsearch/indices/SystemIndexManagerIT.java

```diff
@@ -0,0 +1,255 @@
+ */
+
+package org.elasticsearch.indices;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.admin.indices.mapping.get.GetMappingsRequest;
+import org.elasticsearch.action.admin.indices.mapping.get.GetMappingsResponse;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.MappingMetadata;
+import org.elasticsearch.cluster.service.ClusterService;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.plugins.SystemIndexPlugin;
+import org.elasticsearch.test.ESIntegTestCase;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.elasticsearch.test.XContentTestUtils.convertToXContent;
+import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
+import static org.hamcrest.Matchers.equalTo;
+
+@ESIntegTestCase.ClusterScope(scope = ESIntegTestCase.Scope.TEST, numDataNodes = 0)
+public class SystemIndexManagerIT extends ESIntegTestCase {
+
+    private static final String INDEX_NAME = ".test-index";
+    private static final String PRIMARY_INDEX_NAME = INDEX_NAME + "-1";
+
+    @Before
+    public void beforeEach() {
+        TestSystemIndexDescriptor.useNewMappings.set(false);
+    }
+
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return CollectionUtils.appendToCopy(super.nodePlugins(), TestPlugin.class);
+    }
+
+    /**
+     * Check that if the the SystemIndexManager finds a managed index with out-of-date mappings, then
+     * the manager updates those mappings.
+     */
+    public void testSystemIndexManagerUpgradesMappings() throws Exception {
+        internalCluster().startNodes(1);
+
+        // Trigger the creation of the system index
+        assertAcked(prepareCreate(INDEX_NAME));
+        ensureGreen(INDEX_NAME);
+
+        assertMappings(TestSystemIndexDescriptor.getOldMappings());
+
+        // Poke the test descriptor so that the mappings are now "updated"
+        TestSystemIndexDescriptor.useNewMappings.set(true);
+
+        // Cause a cluster state update, so that the SystemIndexManager will update the mappings in our index
+        triggerClusterStateUpdates();
+        waitNoPendingTasksOnAll(); // FIXME why do I need this to see the new mappings?
```

Comment:
Maybe just do the assertMappings in an assertBusy because the update is async and not controlled by the test? Then you can remove this line